File: sunflower-core/src/main/java/org/tdf/sunflower/sync/GetBlocks.java
package org.tdf.sunflower.sync;

public class GetBlocks {
    private long startHeight;
    private long stopHeight;
    private boolean descend;
    private int limit;

    public GetBlocks(long startHeight, long stopHeight, boolean descend, int limit) {
        this.startHeight = startHeight;
        this.stopHeight = stopHeight;
        this.descend = descend;
        this.limit = limit;
    }

    public GetBlocks() {
    }

    GetBlocks clip() {
        if (stopHeight - startHeight < limit) return this;
        if (descend) {
            startHeight = stopHeight - limit;
        } else {
            stopHeight = startHeight + limit;
        }
        return this;
    }

    public long getStartHeight() {
        return this.startHeight;
    }

    public long getStopHeight() {
        return this.stopHeight;
    }

    public boolean isDescend() {
        return this.descend;
    }

    public int getLimit() {
        return this.limit;
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/sync/GetBlocks.kt
package org.tdf.sunflower.sync

import com.github.salpadding.rlpstream.annotation.RlpCreator
import com.github.salpadding.rlpstream.annotation.RlpProps

@RlpProps("startHeight", "stopHeight", "descend", "limit")
data class GetBlocks @RlpCreator constructor(
    val startHeight: Long,
    val stopHeight: Long,
    val descend: Boolean,
    val limit: Int
) {
    fun clip(): GetBlocks {
        var startHeight = this.startHeight
        var stopHeight = this.stopHeight
        if (stopHeight - startHeight < limit) return this.copy()
        if (descend) {
            startHeight = this.stopHeight - limit
        } else {
            stopHeight = this.startHeight + limit
        }
        return GetBlocks(startHeight, stopHeight, descend, limit)
    }
}